restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/SmsService.java

```diff
@@ -169,19 +171,64 @@ private void message(final Object message) throws IOException {
         final SipURI fromURI = (SipURI) request.getFrom().getURI();
         final String fromUser = fromURI.getUser();
         final ClientsDao clients = storage.getClientsDao();
+
+        final SipURI ruri = (SipURI) request.getRequestURI();
+        final String to = ruri.getUser();
+
+        Sid destinationOrganizationSid = SIPOrganizationUtil.searchOrganizationBySIPRequest(storage.getOrganizationsDao(), request);
+
         final Sid sourceOrganizationSid = OrganizationUtil.getOrganizationSidBySipURIHost(storage, fromURI);
+
         if(logger.isDebugEnabled()) {
             logger.debug("sourceOrganizationSid: " + sourceOrganizationSid);
+            logger.debug("destinationOrganizationSid: "+destinationOrganizationSid);
         }
         if(sourceOrganizationSid == null){
             logger.error("Null Organization: fromUri: "+fromURI);
         }
+
         final Client client = clients.getClient(fromUser, sourceOrganizationSid);
+        final String toUser = CallControlHelper.getUserSipId(request, useTo);
+        final Client toClient = clients.getClient(toUser, destinationOrganizationSid);
         final AccountsDao accounts = storage.getAccountsDao();
         final ApplicationsDao applications = storage.getApplicationsDao();
 
+        IncomingPhoneNumber number = getIncomingPhoneNumber(to, sourceOrganizationSid, destinationOrganizationSid);
+
+        if (number != null) {
+            Account numAccount = accounts.getAccount(number.getAccountSid());
+
+            if (!numAccount.getStatus().equals(Account.Status.ACTIVE)) {
+                //reject SMS since the Number belongs to an an account which is not ACTIVE
+                final SipServletResponse response = request.createResponse(SC_NOT_ACCEPTABLE);
+                response.send();
+
+                String msg = String.format("Restcomm rejects this SMS because number's %s account %s is not ACTIVE, current state %s", number.getPhoneNumber(), numAccount.getSid(), numAccount.getStatus());
+                if (logger.isDebugEnabled()) {
+                    logger.debug(msg);
+                }
+                sendNotification(msg, 11005, "error", true);
+                return;
+            }
+        }
+
         // Make sure we force clients to authenticate.
         if (client != null) {
+
+            Account clientAccount = accounts.getAccount(client.getAccountSid());
+            if (!clientAccount.getStatus().equals(Account.Status.ACTIVE)) {
+                //reject SMS since the Number belongs to an an account which is not ACTIVE
+                final SipServletResponse response = request.createResponse(SC_NOT_ACCEPTABLE);
```

Comment:
any reason to use 406/NOT_ACCEPTABLE? AFAIK this is more related to Accep header and contentNegotiation. what about 403/FORBIDDEN? 403 is a general purpose status to reject requests "The server understood the request, but is refusing to fulfill it" (doesnt have to be related to authorization alone). 

Also consider to put a reason phrase in addition to the code https://docs.oracle.com/cd/E13209_01/wlcp/wlss40/javadoc/jsr289/index.html?javax/servlet/sip/SipServlet.html like "Account is not Active"

Reply:
@jaimecasero No objection on the 403/Forbidden but I think we shouldn't provide reason phrase for security reasons, I think RC shouldn't provide hints that the account is not activated